apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.cs

```diff
@@ -108,7 +113,19 @@ public static UrlSigner FromServiceAccountCredential(ServiceAccountCredential cr
         public static UrlSigner FromBlobSigner(IBlobSigner signer)
         {
             GaxPreconditions.CheckNotNull(signer, nameof(signer));
-            return new UrlSigner(signer);
+            return new UrlSigner(signer, SigningVersion.Default);
+        }
+
+        /// <summary>
+        /// Returns a new instance of <see cref="UrlSigner"/> using the same signing source (e.g. a service account
+        /// private key) but with the specified URL signing algorithm version.
+        /// </summary>
+        /// <param name="signingVersion">The algorithm version to use when signing URLs.</param>
+        /// <returns>A new instance using the specified version.</returns>
+        public UrlSigner WithSigningVersion(SigningVersion signingVersion)
+        {
+            GaxPreconditions.CheckEnumValue(signingVersion, nameof(signingVersion));
+            return new UrlSigner(_blobSigner, signingVersion);
```

Comment:
Nearly. Just that statement ignores the return value, whereas you'll want to use it:

```csharp
var signer = UrlSigner.FromServiceAccountCredential(credentials).WithSigningVersion(SigningVersion.V4);
var url = signer.Sign(/* normal arguments */);
```

The aim would be that if a user just used `UrlSigner.FromServiceAccountCredential(...)` then at the moment they'll get V2, and in a later release they'll get "V4 unless the expiry time makes that infeasible".

Reply:
Thanks @jskeet!
1. W.r.t to your general comment: `If the signature part varies based on version (not just what gets signed, but how it's signed) that won't work any more`, are you only referring to the RSA signer or the signature format as well?
1. When you say "V4 unless the expiry time makes that infeasible", do you mean it will fall back to v2? If so, the client shouldn't do that and instead fail on expiration time longer than a week.

Reply:
1. Unsure at the moment until I've looked in more detail. This is an area I'm only moderately familiar with, and I don't want to make it sound like I know more than I actually do.
2. I'd prefer the default to fall back, to avoid this being a breaking change. Unless we bump the major version number, clients should expect to be able to upgrade to a new version without their code breaking. (If they're expecting a very specific format of URL, it's reasonable for them to have to change - but it feels like "Does what it says on the tin" should keep working.)

Reply:
1. Understood, PLMK when you dive in I'm curious :smile:.
2. W.r.t @theacodes's proposal, client libraries would add a warning when using V2 as the default stating that eventually it will default to V4. "Eventually" is still uncertain at this time. Keeping developers who use these methods in the loop when making the change in a reasonable time along with migration documentation will help support not needing to release a major version.

Reply:
We don't really have a warning option in .NET. I suspect .NET developers are also more wary of breaking changes than some other platforms. We can dive into this more internally.

Reply:
Sg, thanks @jskeet!

Reply:
Oh, btw, given the surface design right now. I think this concept does fit the expectation and the experience is clear to me.

Given the caveat that you will still need to follow-up on changes in implementation and how that may impact this design as well.